network-api/networkapi/wagtailpages/pagemodels/products.py

```diff
@@ -426,6 +480,86 @@ def get_context(self, request, *args, **kwargs):
+                    product.current_total = product.current_total + value
+                    product.current_votecount = product.current_votecount + 1
+
+                    # Add the vote to the vote bin
+                    if not product.vote_bin:
+                        # If there is no vote bin attached to this product yet, create one now.
+                        vote_bin = ProductPageVoteBin()
+                        vote_bin.save()
+                        product.vote_bin = vote_bin
+
+                    # Add the vote to the proper "vote bin"
+                    if value <= 20:
+                        product.vote_bin.bin_1 = product.vote_bin.bin_1 + 1
+                    elif value <= 40:
+                        product.vote_bin.bin_2 = product.vote_bin.bin_2 + 1
+                    elif value <= 60:
+                        product.vote_bin.bin_3 = product.vote_bin.bin_3 + 1
+                    elif value <= 80:
+                        product.vote_bin.bin_4 = product.vote_bin.bin_4 + 1
+                    elif value <= 100:
+                        product.vote_bin.bin_5 = product.vote_bin.bin_5 + 1
+
+                    # Save the product
+                    product.vote_bin.save()
+                    # Don't save this as a revision with .save_revision() as to not spam the Audit log
+                    # And don't make this live with .publish(). The Page model will have the proper
+                    # data stored on it already, and the revision history won't be spammed by votes.
+                    product.save()
+                    return HttpResponse('Vote recorded', status=201, content_type='text/plain')
+                except ProductPage.DoesNotExist:
+                    return HttpResponse('Missing page', status=400, content_type='text/plain')
+                except ValidationError as ex:
+                    return HttpResponse(f'Payload validation failed: {ex}', status=400, content_type='text/plain')
+                except Error as ex:
+                    print(f'{ex.message} ({type(ex)})')
+                    return HttpResponse('Internal Server Error', status=500, content_type='text/plain')
+        elif not self.vote_bin:
+            # Double check a voting bin exists. It should always exist.
+            # TODO: Test the Product-to-ProductPage migration to ensure vote_bin always exists.
+            # If all vote bins exist, we can safely remove this elif statement.
+            vote_bin = ProductPageVoteBin()
+            vote_bin.save()
+            self.vote_bin = vote_bin
+            self.save()
+
+        return super().serve(request, *args, **kwargs)
+
+    def save(self, *args, **kwargs):
+        # When a new ProductPage is created, ensure a vote bin always exists.
+        # We can use save() or a post-save Wagtail hook.
+        if not self.vote_bin:
+            vote_bin = ProductPageVoteBin()
```

Comment:
`ProductPageVoteBin`'s wont be automagically created when a page is saved. We're making sure one _always_ exists when the page is saved. This ensures future ProductPages added will have the voting bin it needs and prevents the page from 500ing and throwing an issue in the template view